src/inference/involution_dsl.jl

```diff
@@ -0,0 +1,440 @@
+import ForwardDiff
+import MacroTools
+import LinearAlgebra
+
+# TODO allow providing a pair of address namespaces, one for the model and one
+# for the proposal, when calling another involution function with call
+
+# TODO proposal_retval is a way that continuous data can leak unaccounted for..
+# for now, the requirement is that the return value of f_disc cannot depend on
+# continuous addresses in the model or proposal although in the future, we
+# could do AD through its return value as well using choice_gradients() and add
+# these to the Jacobian...?
+
+# TODO add more checks that we don't visit the same address twice (these can be
+# only enabled in 'check' mode)
+
+struct InvolutionDSLProgram
+    fn!::Function
+end
+
+struct FirstPassState
+    trace
+    u::ChoiceMap
+    constraints::ChoiceMap
+    u_back::ChoiceMap
+    t_cont_reads::Dict
+    u_cont_reads::Dict
+    t_cont_writes::Dict
+    u_cont_writes::Dict
+    t_copy_reads::DynamicSelection
+    u_copy_reads::DynamicSelection
+    marked_as_retained::Set
+end
+
+function FirstPassState(trace, u::ChoiceMap)
+    FirstPassState(
+        trace, u, choicemap(), choicemap(),
+        Dict(), Dict(), Dict(), Dict(),
+        DynamicSelection(), DynamicSelection(),
+        Set())
+end
+
+struct JacobianPassState{T}
+    trace
+    u::ChoiceMap
+    input_arr::AbstractArray{T}
+    output_arr::Array{T}
+    t_key_to_index::Dict
+    u_key_to_index::Dict
+    cont_constraints_key_to_index::Dict
+    cont_u_back_key_to_index::Dict
+end
+
+const inv_state = gensym("inv_state")
+
+"""
+    @involution function f(...)
+        ..
+    end
+
+Write a program in the [Involution DSL](@ref).
+"""
+macro involution(ex)
+    MacroTools.@capture(ex, function f_(args__) body_ end) || error("expected syntax: function f(..) .. end")
```

Comment:
Not too important -- but I think `MacroTools` has plug-and-play utilities for matching function definitions no matter how they are phrased (Xuan recently changed `@gen` to use them, and the `@dist` DSL does too). Docs here: http://mikeinnes.github.io/MacroTools.jl/stable/utilities/#Function-definitions-1

It might be good to make the one-line change here, so that people could write `@involution f(...) = begin ... end` too (if that happens to be their style).

Reply:
+1 to using the `MacroTools` util, but @alex-lew I gotta say, if you come up with a case where someone can reasonably define their involution in one line, let me know.

Reply:
Done